hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/update/server/SCMUpdateClientManager.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.UpdateRequest;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.UpdateResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Class that manages SCM update clients.
+ */
+public class SCMUpdateClientManager {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SCMUpdateClientManager.class);
+  private Map<UUID, SCMUpdateClientInfo> clients;
+  private Map<Type, SCMUpdateHandler> handlers;
+
+  public SCMUpdateClientManager() {
+    clients = new ConcurrentHashMap<>();
+    handlers = new ConcurrentHashMap<>();
+  }
+
+  public void registerHandler(SCMUpdateHandler handler) {
+    handlers.put(handler.getType(), handler);
+  }
+
+  public void unRegisterHandler(Type type) {
+    handlers.remove(type);
+  }
+
+  public UUID addClient() throws SCMException {
+    UUID clientId = UUID.randomUUID();
+    int retryCount = 5;
+    while (clients.containsKey(clientId)) {
+      if (retryCount > 0) {
+        clientId = UUID.randomUUID();
+        retryCount--;
+      } else {
+        throw new SCMException("Failed to add CRL client with random clientId" +
+            " collision", SCMException.ResultCodes.FAILED_TO_ADD_CRL_CLIENT);
+      }
+    }
+
+    SCMUpdateClientInfo clientInfo = new SCMUpdateClientInfo(clientId);
+    clients.put(clientId, clientInfo);
+    return clientId;
+  }
+
+  // this does not necessarily produce a server response via responseObserver.
+  public void handleClientUpdate(UpdateRequest request,
+      StreamObserver<UpdateResponse> responseObserver) {
+    UUID clientId = SCMUpdateClientInfo.fromClientIdProto(
+        request.getClientId());
+
+    // Unknown client update
+    if (!clients.containsKey(clientId)) {
+      responseObserver.onError(Status.INVALID_ARGUMENT
+          .withDescription("Client must subscribe before it can " +
+              "send/receive updates")
+          .asException());
+    }
+
+    // record the server to client channel
+    SCMUpdateClientInfo clientInfo = clients.get(clientId);
+    if (clientInfo.getResponseObserver() == null) {
+      clientInfo.setResponseObserver(responseObserver);
+    }
+
+    if (handlers.containsKey(request.getUpdateType())) {
+      handlers.get(request.getUpdateType())
+          .handleClientRequest(request, clientInfo);
+    } else {
+      responseObserver.onError(Status.INVALID_ARGUMENT
+          .withDescription("Unknown client update type.")
+          .asException());
+    }
+  }
+
+  /**
+   * Remove client by client Id.
+   * @param clientId - client Id
+   * @return true if client is removed, false otherwise.
+   */
+  public boolean removeClient(UUID clientId) {
+    if (clients.containsKey(clientId)) {
+      SCMUpdateClientInfo clientInfo = clients.remove(clientId);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `remove()` may not be thread-safe at lines: **110 and 111**. Two threads can perform this **same check at the same time** and the second thread to call `remove()` will get `null`. This `null` may be used at line: **112**.

**Fix**  
 Consider calling `remove()`, checking instead of your current check if the returned object is `null`, and only then using that object.